pkg/distributor/ha_tracker.go

```diff
@@ -312,6 +319,7 @@ func (h *haTracker) updateKVStoreAll(ctx context.Context, now time.Time) {
 			} else if h.withinUpdateTimeout(now, entry.nonElectedLastSeenTimestamp) {
 				// Not seen elected but have seen another: attempt to fail over.
 				replica = entry.nonElectedLastSeenReplica
+				level.Info(h.logger).Log("msg", "new replica elected after update timeout exceeded", "user", userID, "cluster", cluster, "previous_replica", entry.elected.Replica, "elected_replica", replica)
```

Comment:
This is where we might try to elect a new replica, but we may not succeed.

I would add logs in two places:
- In the CAS operation, stating our intention to update the replica (no need to state whether it succeeded)
- In the updateCache operation, stating the update effect (from this or other distributors)

Like this:
```diff
diff --git pkg/distributor/ha_tracker.go pkg/distributor/ha_tracker.go
index 12d558bfc..823fcdc6a 100644
--- pkg/distributor/ha_tracker.go
+++ pkg/distributor/ha_tracker.go
@@ -147,7 +147,7 @@ func newHATracker(cfg HATrackerConfig, limits haTrackerLimits, reg prometheus.Re
        }

        t := &haTracker{
-               logger:              logger,
+               logger:              log.With(logger, "component", "ha-tracker"),
                cfg:                 cfg,
                updateTimeoutJitter: jitter,
                limits:              limits,
@@ -462,6 +462,7 @@ func (h *haTracker) updateCache(userID, cluster string, desc *ReplicaDesc) {
        }
        if desc.Replica != entry.elected.Replica {
                h.electedReplicaChanges.WithLabelValues(userID, cluster).Inc()
+               level.Info(h.logger).Log("msg", "updating replica in cache", "user", userID, "cluster", cluster, "old_replica", entry.elected.Replica, "new_replica", desc.Replica, "received_at", timestamp.Time(desc.ReceivedAt))
        }
        entry.elected = *desc
        h.electedReplicaTimestamp.WithLabelValues(userID, cluster).Set(float64(desc.ReceivedAt / 1000))
@@ -476,11 +477,18 @@ func (h *haTracker) updateKVStore(ctx context.Context, userID, cluster, replica
                var ok bool
                if desc, ok = in.(*ReplicaDesc); ok && desc.DeletedAt == 0 {
                        // If the entry in KVStore is up-to-date, just stop the loop.
-                       if h.withinUpdateTimeout(now, desc.ReceivedAt) ||
-                               // If our replica is different, wait until the failover time.
-                               desc.Replica != replica && now.Sub(timestamp.Time(desc.ReceivedAt)) < h.cfg.FailoverTimeout {
+                       if h.withinUpdateTimeout(now, desc.ReceivedAt) {
                                return nil, false, nil
                        }
+
+                       // If our replica is different, wait until the failover time.
+                       if desc.Replica != replica {
+                               if now.Sub(timestamp.Time(desc.ReceivedAt)) < h.cfg.FailoverTimeout {
+                                       level.Info(h.logger).Log("msg", "replica differs, but it's too early to failover", "user", userID, "cluster", cluster, "replica", replica, "elected", desc.Replica, "received_at", timestamp.Time(desc.ReceivedAt))
+                                       return nil, false, nil
+                               }
+                               level.Info(h.logger).Log("msg", "replica differs, attempting to update kv", "user", userID, "cluster", cluster, "replica", replica, "elected", desc.Replica, "received_at", timestamp.Time(desc.ReceivedAt))
+                       }
                }

                // Attempt to update KVStore to our timestamp and replica.
```